# MONTECARLO PARALELIZADO

## Instalaciones necesarias

- Para que el código a continuación funcione, es necesario tener instalado `dask`.

In [1]:
%pip install --user dask   

     |████████████████████████████████| 798kB 606kB/s eta 0:00:01
You should consider upgrading via the 'pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [2]:
%pip install --user dask distributed

     |████████████████████████████████| 614kB 598kB/s eta 0:00:01
     |████████████████████████████████| 276kB 4.4MB/s eta 0:00:01
     |████████████████████████████████| 276kB 2.4MB/s eta 0:00:01
     |████████████████████████████████| 450kB 2.5MB/s eta 0:00:01
     |████████████████████████████████| 51kB 1.0MB/s eta 0:00:011
     |████████████████████████████████| 92kB 1.6MB/s eta 0:00:011
  Created wheel for pyyaml: filename=PyYAML-5.3.1-cp36-cp36m-linux_x86_64.whl size=45919 sha256=4dcb8da69c2beb3982c2bb80d57f5b83c3226b2ae2b7918e05b1dba59649b6ff
  Stored in directory: /home/miuser/.cache/pip/wheels/a7/c1/ea/cf5bd31012e735dc1dfea3131a2d5eae7978b251083d6247bd
  Created wheel for psutil: filename=psutil-5.7.0-cp36-cp36m-linux_x86_64.whl size=279874 sha256=01c5228e1ad4ddfa5acd50eea4984a382ff057b50022f95c2ec08ea7cadbc497
  Stored in directory: /home/miuser/.cache/pip/wheels/d7/69/b4/3200b95828d1f0ddb3cb5699083717f4fdbd9b4223d0644c57
  Created wheel for toolz: filename=toolz-0.10.0-cp36

- Para habilitar el dashboard, debemos installar `bokeh` :

In [3]:
%pip install --user bokeh

     |████████████████████████████████| 8.6MB 4.2MB/s eta 0:00:01
     |████████████████████████████████| 20.2MB 5.2MB/s eta 0:00:01
     |████████████████████████████████| 2.1MB 548kB/s eta 0:00:01
     |████████████████████████████████| 71kB 1.3MB/s eta 0:00:011
  Created wheel for bokeh: filename=bokeh-2.0.1-cp36-none-any.whl size=9082987 sha256=b813f5f1513d4ba1b413abc68fdd20332a1cb78ceedb58d9ffce4c88cf69119c
  Stored in directory: /home/miuser/.cache/pip/wheels/de/64/d0/e7f70919e19f57a6efceb65cc4edaa0449e707bcdb6e7e61ad
Successfully built bokeh
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
You should consider upgrading via the 'pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [4]:
%%bash 
sudo apt-get install -y graphviz

Reading package lists...
Building dependency tree...
Reading state information...
The following additional packages will be installed:
  fontconfig fontconfig-config fonts-dejavu-core fonts-liberation libann0
  libcairo2 libcdt5 libcgraph6 libdatrie1 libfontconfig1 libfreetype6 libgd3
  libgraphite2-3 libgts-0.7-5 libgts-bin libgvc6 libgvpr2 libharfbuzz0b
  libice6 libjbig0 libjpeg-turbo8 libjpeg8 liblab-gamut1 libltdl7
  libpango-1.0-0 libpangocairo-1.0-0 libpangoft2-1.0-0 libpathplan4
  libpixman-1-0 libpng16-16 libsm6 libthai-data libthai0 libtiff5 libwebp6
  libxaw7 libxcb-render0 libxcb-shm0 libxmu6 libxpm4 libxrender1 libxt6 ucf
  x11-common
Suggested packages:
  gsfonts graphviz-doc libgd-tools
The following NEW packages will be installed:
  fontconfig fontconfig-config fonts-dejavu-core fonts-liberation graphviz
  libann0 libcairo2 libcdt5 libcgraph6 libdatrie1 libfontconfig1 libfreetype6
  libgd3 libgraphite2-3 libgts-0.7-5 libgts-bin libgvc6 libgvpr2 libharfbuzz0b
  libice6 l

E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/libg/libgd2/libgd3_2.2.5-4ubuntu0.3_amd64.deb  404  Not Found [IP: 91.189.88.142 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?


CalledProcessError: Command 'b'sudo apt-get install -y graphviz\n'' returned non-zero exit status 100.

In [6]:
%pip install --user -q graphviz

You should consider upgrading via the 'pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [5]:
%pip install --user numpy 

You should consider upgrading via the 'pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [7]:
import IPython
app=IPython.Application.instance()
app.kernel.do_shutdown(True) #restarting kernell

{'status': 'ok', 'restart': True}

## Montecarlo 1D


Empezamos con el algoritmo de monte carlo en 1 dimensión

Integramos la función:
$$f(x)=\int_0^1 \frac{4}{1+x^2}=\pi$$ 




### Método secuencial:

In [1]:
import numpy as np
import math

In [2]:
#definimos la funcion del error relativo
err_rel=lambda ap,ob: math.fabs(ap-ob)/math.fabs(ob)

In [3]:
#numero de puntos a utilizar
density_p=10**8

#funcion a integrar
f=lambda x: 4/(1+x**2)

#resultado conocido de la integral (valor objetivo)
obj=math.pi

#intervalo
a=0
b=1

#dimensiones
dims=1

In [4]:
%%time
#approximación mediante método montecarlo
x_p=np.random.uniform(0,1,density_p) #obtiene tantos números aleatorios entre 0 y 1 como density_p.
vol=b-a # esta es la longitud del intervalo de integración.
approx=vol*np.mean(f(x_p))

print("---------------------")
print("El valor calculado de la integral es: ",approx)
print("El error relativo: {:0.4e}".format(err_rel(approx,obj)))
print("---------------------")

---------------------
El valor calculado de la integral es:  3.141591206171679
El error relativo: 4.6073e-07
---------------------
CPU times: user 912 ms, sys: 273 ms, total: 1.19 s
Wall time: 1.18 s


### Paralelización de MC en Dask

Para paralelizar estos resultados lo que pienso que tenemos que hacer es dividir las tareas en n subintervalos. La porpuesta es la siguiente:  
1. Dividir el dominio de la función en $p$ partes iguales. $p$ será el número de cpus disponibles en la máquina en que se corre la tarea.
2. Dividir el total de puntos a utilizar equitativamente entre $p$.
3. Resolver el método para cada intervalo.
4. Sumar los resultados de cada intervalo.

Nota: Para el caso de más dimensiones solamente se paraleliza una dimensión pues al hacerlo con todas se estaría dividiendo el espacio en $p_{dim}$ subespacios.

In [15]:
import multiprocessing
import time
from dask.distributed import Client
client = Client()

In [16]:
client

<Client: 'tcp://127.0.0.1:33357' processes=4 threads=8, memory=8.21 GB>

In [17]:
#paso 1. Dividir el dominio en partes iguales
p=multiprocessing.cpu_count() #cpus disponibles
n_subint=int(density_p/p) #numero de puntos o nodos en cada core o cpu

In [18]:
def construye_subintervalos1D(ids,a,b,p,dims):
    """

    Función que construye "p" subintervalos, delimitados entre a y b. El número de subintervalos \\
    dependerá del número de cores empleados. Este ejercicio fue ejecutado en una computadora \\
    con 16 cores, por lo tanto, se crean 16 subintervalos.
    
    Argumentos:
    ----------
    * ids: Identificador del core dónde se está corriendo el task. Este ejercicio fue ejecutado \\
           en una computadora con 16 cores, por ello los posibles valores de "ids" son números enteros \\
           entre 0 y 15.
    * a, b (float): Intervalo de integración [a, b]
    * p (int) : Número de cores o cpus disponibles
    * dims: Dimensión de la integral.
    
    Salidas:
    -------
    * (begin, end): Subintervalos de integración.
    
    Ejemplo:
    -------
    construye_subintervalos(ids=0,a,b,p,dims)
    return:
    (0.0, 0.0625)

    """
    tamano_int=(b-a)/p #tamaño de cada sub intervalo.
    begin= a + ids*tamano_int #construyen los subintervalo
    end=begin+tamano_int
    return (begin,end)

    
def evalua_subintervalos1D(intervalo,f,n_subint,dims):
    """
    Función que evalúa la integral en cada uno los p subintervalos \\
    previamente construidos con la función "construye_subintervalos"
    
    Argumentos:
    ----------
    * intervalo: Intervalo de la integral sobre la cual se va aproximar \\
                 la integral
    * f: Función que se está aproximando
    * n_subint: Número de puntos o nodos en cada core o cpu
    * dims: Dimensiones de la integral
    
    Salidas:
    --------
    + vol*np.mean(f(x_p)): Approximación a la integral
    
    """   
    x_p=np.random.uniform(intervalo[0],intervalo[1],n_subint) #draw samples from a uniform distribution
    vol=intervalo[1]-intervalo[0]
    return vol*np.mean(f(x_p)) #calcula la integral para una cada regió: subintervalo.

In [19]:
%%time

#submitting p function calls
futures_intervalos = client.map(construye_subintervalos1D,range(p),
                                                **{'a':a,
                                                   'b':b,
                                                   'p':p,
                                                   'dims':dims})

futures_evaluando=client.map(evalua_subintervalos1D, futures_intervalos,
                                             **{'f':f,
                                                'n_subint':n_subint,
                                                'dims':dims})

results=client.gather(futures_evaluando) # los resultados viven en varios workers 
aprox=sum(results) #sumamos los resultados de todos los workers

print("---------------------")
print("El valor calculado de la integral es: ",aprox)
print("El error relativo: {:0.4e}".format(err_rel(aprox,obj)))
print("---------------------")

---------------------
El valor calculado de la integral es:  3.1415892139239507
El error relativo: 1.0949e-06
---------------------
CPU times: user 182 ms, sys: 9.23 ms, total: 192 ms
Wall time: 891 ms


In [20]:
client.close()

Para el método en secuencial obtenemos un error relativo de $3.6251x10^{-6}$ y un tiempo de ejecución de $2.38$ seg y para el método en paralelo obtenemos un error relativo de $6.0184x10^{-7}$ en $110$ ms.

# Monte carlo en n-dimensiones

Se hace la generalización en que solamente se divide la primera dimensión y se paraleliza sobre esa dimensión usando las otras diemnsiones completas.  
Los intervalos sobre los que se trabaja son intervalos divididos en $p$ partes iguales para la primera dimensión y el intervalo de intergración completo para cada una de las siguientes dimensiones.

In [21]:
#funciones para multiples dimensiones

def construye_subintervalos(ids,a,b,p,dims):
    """
    Función que construye "p" subintervalos, delimitados entre a y b para la primera dimensión.    
    En las demás dimensiones integra sobre el intervalo completo.
    
    Argumentos:
    ----------
    * ids: Identificador del core dónde se está corriendo el task. 
    * a (float): Vector con el límite inferior de cada intervalo de integración [ax1,ax2,...,axn] para cada diemnsión. Tiene dims entradas.
    * b (float): Vector con el límite superior de cada intervalo de integración [bx1,bx2,...,bxn] para cada diemnsión. Tiene dims entradas.
    * p (int) : Número de cores o cpus disponibles
    * dims: Dimensiónes de la integral.
    
    Salidas:
    -------
    * intervalos : lista de pares de intervalos de integración ordenados por dimensión
    
    """
    if(dims==1):
        tamano_int=(b-a)/p
        begin= a + ids*tamano_int
        end=begin+tamano_int
        return (begin,end)

    else:
        tamano_int=(b-a)/p

        #aux solamente cambia la primera dimension
        aux=np.identity(dims)
        aux=aux[0,:]
        aux=aux*ids*tamano_int
        begin= a + aux
        
        aux=np.identity(dims)
        aux=aux[0,:]
        aux=aux*tamano_int
        end=begin+aux
        
        intervalos=[]
        for i in range(dims):
            if(i==0):
                intervalos=np.append(intervalos,[begin[i],end[i]])
            else:
                intervalos=np.append(intervalos,[a[i],b[i]])
                
        return intervalos


In [23]:
def evalua_subintervalos(intervalo,f,n_subint,dims):
    """
    Función que evalúa la integral en cada uno los p subintervalos \\
    previamente construidos con la función "construye_subintervalos"
    
    Argumentos:
    ----------
    * intervalo: Intervalo de la integral sobre la cual se va aproximar \\
                 la integral
    * f: Función que se está aproximando
    * n_subint: Número de puntos o nodos en cada core o cpu
    * dims: Dimensiones de la integral
    
    Salidas:
    --------
    + vol*np.mean(f(x)): Approximación a la integral
    
    """   

    x=[]
    vol=1
    for i in range(dims):
        x=np.append(x,np.random.uniform(intervalo[2*i],intervalo[2*i+1],n_subint))
        vol=vol*(intervalo[2*i+1]-intervalo[2*i])
    
    x=np.reshape(x,(n_subint,dims), order='F')
    return np.mean(f(x))*vol

### Ejemplo 1
 $ \int_0^2 \int_{-1}^1 \int_0^1 (2x+3y+z)dzdydx = 10$
 
metodo en paralelo

In [24]:
client=Client()

#numero de procesadores
p=multiprocessing.cpu_count()

#numero de puntos a utlizar
density_p=10**7
n_subint=int(density_p/p)

#funcion a integrar
# cada variable se toma como la columna de una matriz
f=lambda x: 2*x[:,0]+3*x[:,1]+x[:,2]

#reultado conocido de la integral (valor objetivo)
obj=10

#intervalo
# se debe de poner (x1_a,x2_a,...,xn_a) para los mínimos
a=np.array([0,-1,0])
# se debe de poner (x1_b,x2_b,...,xn_b) para los máxios
b=np.array([2,1,1])

#volumen
vol=np.prod(b-a)

#dimensiones
dims=len(a)

In [27]:
%%time
#llamado para multiples dimensiones

futures_intervalos = client.map(construye_subintervalos, range(p),
                                            **{'a':a,
                                               'b':b,
                                               'p':p,
                                               'dims':dims})

futures_evaluando=client.map(evalua_subintervalos, futures_intervalos,
                                            **{'f':f,
                                               'n_subint':n_subint,
                                               'dims':dims})

results=client.gather(futures_evaluando)
aprox=sum(results)

print("---------------------")
print("El valor calculado de la integral es: ",aprox)
print("El error relativo: {:0.4e}".format(err_rel(aprox,obj)))
print("---------------------")

---------------------
El valor calculado de la integral es:  10.003357009626413
El error relativo: 3.3570e-04
---------------------
CPU times: user 132 ms, sys: 1.69 ms, total: 133 ms
Wall time: 364 ms


In [28]:
%%time
np.random.seed(1992814)
#density_p=10**7

#f=lambda x,y,z: 2*x+3*y+z
f=lambda x: 2*x[:,0]+ 3*x[:,1]+ x[:,2]
density_p=10**7
x0=0
x1=2
y0=-1
y1=1
z0=0
z1=1

x=[]
x=np.append(x,np.random.uniform(x0,x1,density_p))
x=np.append(x,np.random.uniform(y0,y1,density_p))
x=np.append(x,np.random.uniform(z0,z1,density_p))
x=np.reshape(x,(density_p,3),order='F')

obj=10
vol=(x1-x0)*(y1-y0)*(z1-z0)

approx=vol*np.mean(f(x))
err_rel=lambda ap,ob: math.fabs(ap-ob)/math.fabs(ob)
error_est = math.sqrt(sum((f(x)-approx)**2)/density_p)

print("="*50)
print("* Objetivo: ",obj)
print("* Aprox: ",approx)
print("* Error relativo: {:0.4e}".format(err_rel(approx,obj)))
print("* Intervalo: ")
print((approx-vol*error_est, approx+vol*error_est))
print("="*50)


* Objetivo:  10
* Aprox:  9.99794918491403
* Error relativo: 2.0508e-04
* Intervalo: 
(-21.15199619974249, 41.14789456957055)
CPU times: user 3.02 s, sys: 111 ms, total: 3.13 s
Wall time: 3.1 s


Para el método en **secuencial** obtenemos un error relativo de $2.0508x10^{-4}$ y un tiempo de ejecución de $3.13$ seg y para el método en **paralelo** obtenemos un error relativo de $3.357x10^{-4}$ en $133$ ms.